File: 98-validate-binary-search-tree/98-validate-binary-search-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool validater(TreeNode* node, long left, long right){
        
        if(node == NULL) return true;
        
        bool left_tree = validater(node->left, left, node->val);
        bool right_tree = validater(node->right, node->val, right);
        
        if(node->val >= right || node->val <= left) return false; 
        else return (left_tree && right_tree);
    }
    
    bool isValidBST(TreeNode* root) {
        long left = LONG_MIN, right = LONG_MAX;
        
        return validater(root, left, right);
    }
};

File: 98-validate-binary-search-tree/98-validate-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    mini, maxi = -1e11, 1e11
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        
        def checkBST(root, mini, maxi):
            if not root: return True
            
            leftTree = checkBST(root.left, mini, root.val)
            rightTree = checkBST(root.right, root.val, maxi)
            
            if root.val <= mini or root.val >= maxi: return False
            return leftTree and rightTree
        
        return checkBST(root, self.mini, self.maxi)